### PR TITLE
Feature/ispgbss 1177 test for filters and search on component library

### PR DESCRIPTION
**Pull request by rmoore-theta**
*What does this PR do?*

*Jira ticket number?*
[ISPGBSS-1177](https://jiraent.cms.gov/browse/ISPGBSS-1177)

*Changes*

- more integration tests

*Testing*

- If you have all the code up to date, in ui and api repos
- You will also need at least two component types (software and policy) for the component library test to work
- npx cypress run --headless --browser chrome --config video=false

  (Run Finished)

       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  flows/componentLibraryFlow.spec.js       00:35        3        3        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  flows/projectSetupFlow.spec.js           01:02        4        4        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  flows/staticPageFlow.spec.js             00:14        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  general/test.spec.js                     00:15        3        3        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  pages/projectSettingsPage.spec.js        00:09        2        2        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  pages/projectSetupPage.spec.js           00:12        2        2        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  pages/projectSystemCompomnetsMock.s      00:12        5        5        -        -        - │
  │    pec.js                                                                                      │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✔  All specs passed!                        02:43       20       20        -        -        -

### CODE REVIEW

File: src/organisms/Breadcrumbs.jsx [hunk 1]
```diff
@@ -63,22 +63,41 @@ const Breadcrumbs = () => {
...
+      appendToParentList(pathArray[1], `/project-setup/${pathArray[1]}`);
+    }
+    //Adding this one since HCD is working on adding an extra page after selecting components
+    if (pathArray[2] !== undefined && pathArray[3] !== undefined) {
+      appendToParentList(
+        pathArray[2],
+        `/project-setup/${pathArray[1]}/${pathArray[2]}`
+      );
     }
   } else if (pathArray[0] === "projects") {
```
Comment (jauschalley) [inline src/organisms/Breadcrumbs.jsx:78]: I'd much rather see this code broken into testable units with self-documenting names.  We've deprioritized cypress testing due to the effort of getting it into our pipelines.  Unit testing is much easier to add therefore they are more valuable to us at this time.

Comment (rmoore-theta) [reply src/organisms/Breadcrumbs.jsx:78]: Cypress test are easy to run and can do proper integration test in under 3 minutes as pointed out above. They add more value then unit test themselves add, that is why they are important.

File: src/index.js [hunk 2]
```diff
@@ -1,17 +1,15 @@
...
-import { BrowserRouter } from 'react-router-dom';
+import { BrowserRouter } from "react-router-dom";
 import "./index.scss";
-import { AppRoutes } from "./AppRoutes"
+import { AppRoutes } from "./AppRoutes";
 import reportWebVitals from "./reportWebVitals";
 
 const root = ReactDOM.createRoot(document.getElementById("root"));
 root.render(
-  <React.StrictMode>
```
Comment (amyd11) [inline src/index.js]: Hey we shouldn't delete this, it's part of the application that was created in the beginning and it looks like it's a useful tool to continue using (https://reactjs.org/docs/strict-mode.html). Can you add that back in?

File: src/organisms/Breadcrumbs.test.jsx [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+import Breadcrumbs from "./Breadcrumbs";
+
+test("renders breadcrumbs no url", async () => {
+  render(
+    <MemoryRouter initialEntries={[""]}>
+      <App />
+    </MemoryRouter>
+  );
+  await waitFor(() => {
+    const breadcrumbs = screen.queryByTestId("breadcrumbs-bar");
```
Comment (amyd11) [inline src/organisms/Breadcrumbs.test.jsx:16]: Nice, these are great ways to incorporate waiting in tests so that we're not using hardcoded wait values which can produce unreliable test results.

**Review by amyd11 — COMMENTED** [review]
Totally cool with adding cypress tests and I'm not going to block this based on lack of react testing library tests, but we really want to focus on leveraging react tests for the many reasons we've discussed in brainstorming.

I do think we need to add in the react strict mode back in, but once that's fixed I can approve.

File: src/index.js (hunk 2 above)
Comment (rmoore-theta) [reply src/index.js]: Added a ticket to make this dynamic based on environments. This is what causes the double API calls on page load.

File: src/index.js [hunk 4]
```diff
@@ -7,9 +7,12 @@ import reportWebVitals from "./reportWebVitals";
 
 const root = ReactDOM.createRoot(document.getElementById("root"));
 root.render(
-  <BrowserRouter>
-    <AppRoutes />
-  </BrowserRouter>
+  // @TODO: https://jiraent.cms.gov/browse/ISPGBSS-1209
```
Comment (amyd11) [inline src/index.js]: Hmm I didn't know a lot about react strict mode before your PR! But I did some reading about it. I didn't know that it would make multiple api calls and such to help detect state and memory leaks in the app! (Cool stuff) So it's nice that we can use it for development, but obviously don't want it for use in production. Thankfully, react knows this and added a helpful note at the top of their [docs about strict mode](https://reactjs.org/docs/strict-mode.html): 

> **Note:**
Strict mode checks are run in development mode only; _they do not impact the production build._

And when we're ready to deploy to production, we will need to use [a production build](https://reactjs.org/docs/optimizing-performance.html#use-the-production-build) of our app.

**tl;dr** Leaving `<React.StrictMode>` in forever should be safe and we shouldn't have to manually handle removing it for production as long as we correctly do react production builds! :)

File: src/index.js (hunk 2 above)
Comment (amyd11) [reply src/index.js]: Gonna resolve this convo because I left the next comment for this conversation on the TODO line of the code below!

File: src/index.js (hunk 4 above)
Comment (rmoore-theta) [reply src/index.js]: Oh that will make it easier :) -p flag is good to know

File: src/pages/ProjectSetup.test.jsx [hunk 5]
```diff
@@ -38,3 +41,35 @@ test("renders page", () => {
...
+  render(<ProjectSetup />, { wrapper: MemoryRouter });
+  fireEvent.click(screen.getByRole("button", { name: "Next" }));
+  fireEvent.change(screen.getByLabelText("Full name"), {
+    target: { value: "Full Name" },
+  });
+  fireEvent.change(screen.getByLabelText("Acronym"), {
+    target: { value: "Acronym" },
+  });
+  fireEvent.click(screen.getByLabelText("CMS AWS Commercial East-West"));
+  fireEvent.click(screen.getByRole("button", { name: "Next" }));
```
Comment (rmoore-theta) [inline src/pages/ProjectSetup.test.jsx:62]: The reason I keep calling click next is to validate the individual checks that I had to put on each radio button in order for code coverage to increase...

**Review by amyd11 — APPROVED** [review]

File: cypress/integration/flows/componentLibraryFlow.spec.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
+/// <reference types="cypress" />
+
```
Comment (jauschalley) [inline cypress/integration/flows/componentLibraryFlow.spec.js:2]: Praise: Big improvement over the old use of `wait()`

File: src/organisms/Breadcrumbs.jsx [hunk 7]
```diff
@@ -63,36 +65,57 @@ const Breadcrumbs = () => {
...
-
     if (pathArray.length === 2) {
       setPageName(title);
-    } else if (pathArray.length === 3) {
-      //potential pages they are on "Add Team Member", "Project Team", "Project Settings", "System Security Plan", "Manage Components", "System Controls"
+    }
+    if (pathArray[1] !== undefined && pathArray[2] !== undefined) {
       appendToParentList(title, `/projects/${pathArray[1]}`);
     }
+    if (pathArray[2] !== undefined && pathArray[3] !== undefined) {
```
Comment (jauschalley) [inline src/organisms/Breadcrumbs.jsx:90]: Nitpik(non-blocking): This is an example of good place to make a supporting method.  While I can read the code and tell that you're checking for undefined in two specific parts of the pathArray, writing a method that does this, while only used once, gives you two benefits 1) a method name that gives you more context than `pathArray != undefined` and 2) something specific that you can write a very easy unit test for so that the next dev that works on this component doesn't inadvertently break your logic.  In the consulting world, for a single dev to spend two years on a project is a long time so expect a bunch of people from potentially different companies to touch your code after you.

**Review by jauschalley — APPROVED** [review]
Thanks for all of the unit tests.  Just the fact that you added all of these new files for specific components will prompt the devs working on the next iteration to add more tests.
